src/extra/components/pulses.py

```diff
@@ -0,0 +1,422 @@
+"""Interface and utilies to work with pulse-resolved data."""
+
+
+from copy import copy
+from typing import Optional
+import re
+
+import numpy as np
+
+from euxfel_bunch_pattern import PPL_BITS, is_sase, is_laser
+from extra_data import SourceData, KeyData, by_id
+
+
+try:
+    from itertools import pairwise
+except ImportError:
+    # Compatibility for Python < 3.10
+    from itertools import tee
+
+    def pairwise(iterable):
+        # pairwise('ABCDEFG') --> AB BC CD DE EF FG
+        a, b = tee(iterable)
+        next(b, None)
+        return zip(a, b)
+
+
+__all__ = ['PulsePattern']
+
+
+class _PulsePattern:
```

Comment:
I might suggest another property: `is_constant` to detect whether the pulse pattern is constant or not. MID typically uses a constant pulse pattern, but for various reasons the number of pulses in their runs is semi-regularly non-constant, and it could be useful to have a property to check that.

Reply:
Oh, brilliant idea. As it happens, the previous version of the component had that very logic already.

Now I would gravitate towards a method rather than property, is it necessitates a full data read and is thus potentially expensive. Any objection?

Reply:
Nah, method should be fine. If reading is expensive then maybe we could also cache the result, like extra-data does in a lot of places?

Reply:
Ehhh nah that sounds fine.